src/Elasticsearch.Net/Connection/HttpConnection.cs

```diff
@@ -306,10 +307,71 @@ protected virtual HttpRequestMessage CreateRequestMessage(RequestData requestDat
 			return requestMessage;
 		}
 
-		private static void SetContent(HttpRequestMessage message, RequestData requestData) => message.Content = new RequestDataContent(requestData);
+		private static void SetContent(HttpRequestMessage message, RequestData requestData)
+		{
+			if (requestData.TransferEncodingChunked)
+			{
+				message.Content = new RequestDataContent(requestData);
+			}
+			else
+			{
+				var stream = requestData.MemoryStreamFactory.Create();
+				if (requestData.HttpCompression)
+					using (var zipStream = new GZipStream(stream, CompressionMode.Compress, true))
+						requestData.PostData.Write(zipStream, requestData.ConnectionSettings);
+				else
+					requestData.PostData.Write(stream, requestData.ConnectionSettings);
+
+				if (requestData.PostData.DisableDirectStreaming.GetValueOrDefault(false))
+				{
+					message.Content = new ByteArrayContent(requestData.PostData.WrittenBytes);
+					stream.Dispose();
+				}
+				else
+				{
+					stream.Position = 0;
+					message.Content = new StreamContent(stream);
+				}
+
+				if (requestData.HttpCompression)
+					message.Content.Headers.ContentEncoding.Add("gzip");
 
-		private static void SetAsyncContent(HttpRequestMessage message, RequestData requestData, CancellationToken token) =>
-			message.Content = new RequestDataContent(requestData, token);
+				message.Content.Headers.ContentType = new MediaTypeHeaderValue(requestData.RequestMimeType);
+			}
+		}
+
+		private static async Task SetContentAsync(HttpRequestMessage message, RequestData requestData, CancellationToken cancellationToken)
+		{
+			if (requestData.TransferEncodingChunked)
+			{
+				message.Content = new RequestDataContent(requestData, cancellationToken);
+			}
+			else
+			{
+				var stream = requestData.MemoryStreamFactory.Create();
+				if (requestData.HttpCompression)
+					using (var zipStream = new GZipStream(stream, CompressionMode.Compress, true))
+						await requestData.PostData.WriteAsync(zipStream, requestData.ConnectionSettings, cancellationToken).ConfigureAwait(false);
+				else
+					await requestData.PostData.WriteAsync(stream, requestData.ConnectionSettings, cancellationToken).ConfigureAwait(false);
+
+				if (requestData.PostData.DisableDirectStreaming.GetValueOrDefault(false))
+				{
+					message.Content = new ByteArrayContent(requestData.PostData.WrittenBytes);
+					stream.Dispose();
+				}
+				else
+				{
+					stream.Position = 0;
+					message.Content = new StreamContent(stream);
+				}
+
+				if (requestData.HttpCompression)
```

Comment:
There is a similar `if` on line `352` - can they be merged?

Reply:
no they cannot 😃 this `if` sets `gzip` Content Encoding on `message.Content` headers, which is set on line 360 or 366, after line 352.

Reply:
I wonder if this needs a comment to stop the eager refactor! :)

Reply:
If you refactor and then run tests, they will fail - is that reasonable discouragement enough?